File: xy4076/repo/xy4076/podcast_sanitizer/subtitle_parser.py
import re
import csv
from datetime import timedelta
from typing import List, TextIO, Optional
from pathlib import Path

from .models import Subtitle, Chapter, SensitiveRule


def parse_srt_time(time_str: str) -> timedelta:
    time_str = time_str.strip()
    match = re.match(r"(\d+):(\d+):(\d+),(\d+)", time_str)
    if match:
        hours, minutes, seconds, milliseconds = map(int, match.groups())
        return timedelta(
            hours=hours,
            minutes=minutes,
            seconds=seconds,
            milliseconds=milliseconds
        )
    raise ValueError(f"Invalid SRT time format: {time_str}")


def parse_vtt_time(time_str: str) -> timedelta:
    time_str = time_str.strip()
    match = re.match(r"(\d+):(\d+):(\d+)\.(\d+)", time_str)
    if match:
        hours, minutes, seconds, milliseconds = map(int, match.groups())
        return timedelta(
            hours=hours,
            minutes=minutes,
            seconds=seconds,
            milliseconds=milliseconds
        )
    match = re.match(r"(\d+):(\d+)\.(\d+)", time_str)
    if match:
        minutes, seconds, milliseconds = map(int, match.groups())
        return timedelta(
            minutes=minutes,
            seconds=seconds,
            milliseconds=milliseconds
        )
    raise ValueError(f"Invalid VTT time format: {time_str}")


def parse_srt(content: str) -> List[Subtitle]:
    subtitles: List[Subtitle] = []
    blocks = re.split(r"\n\n+", content.strip())

    for block in blocks:
        lines = block.strip().split("\n")
        if len(lines) < 2:
            continue

        try:
            sub_id = int(lines[0].strip())
            time_line = lines[1]

            time_match = re.match(r"(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})", time_line)
            if not time_match:
                continue

            start_time = parse_srt_time(time_match.group(1))
            end_time = parse_srt_time(time_match.group(2))

            text = "\n".join(lines[2:]).strip() if len(lines) > 2 else ""

            subtitles.append(Subtitle(
                id=len(subtitles) + 1,
                start_time=start_time,
                end_time=end_time,
                text=text,
                original_id=sub_id
            ))
        except (ValueError, IndexError):
            continue

    return subtitles


def parse_vtt(content: str) -> List[Subtitle]:
    subtitles: List[Subtitle] = []
    lines = content.split("\n")
    current_id = 1
    i = 0

    while i < len(lines):
        line = lines[i].strip()

        if line.startswith("WEBVTT") or line == "":
            i += 1
            continue

        if "-->" in line:
            time_match = re.match(r"([\d:.]+)\s*-->\s*([\d:.]+)", line)
            if time_match:
                start_time = parse_vtt_time(time_match.group(1))
                end_time = parse_vtt_time(time_match.group(2))

                text_lines = []
                i += 1
                while i < len(lines) and lines[i].strip() != "" and "-->" not in lines[i]:
                    text_lines.append(lines[i].strip())
                    i += 1

                text = "\n".join(text_lines).strip()

                subtitles.append(Subtitle(
                    id=current_id,
                    start_time=start_time,
                    end_time=end_time,
                    text=text
                ))
                current_id += 1
                continue

        i += 1

    return subtitles


def parse_subtitle_file(file_path: str) -> List[Subtitle]:
    path = Path(file_path)
    content = path.read_text(encoding="utf-8")

    if path.suffix.lower() == ".srt":
        return parse_srt(content)
    elif path.suffix.lower() == ".vtt":
        return parse_vtt(content)
    else:
        raise ValueError(f"Unsupported subtitle format: {path.suffix}")


def parse_chapters_csv(content: str) -> List[Chapter]:
    chapters: List[Chapter] = []
    lines = content.split("\n")
    header_skipped = False

    for line in lines:
        line = line.strip()
        if not line:
            continue

        if not header_skipped:
            header_skipped = True
            continue

        parts = line.split(",", 2)
        if len(parts) < 2:
            continue

        try:
            start_time_str = parts[0].strip()
            title = parts[1].strip()

            start_time = parse_vtt_time(start_time_str) if ":" in start_time_str and "." in start_time_str else parse_srt_time(start_time_str)

            end_time = None
            if len(parts) > 2 and parts[2].strip():
                end_time_str = parts[2].strip()
                end_time = parse_vtt_time(end_time_str) if ":" in end_time_str and "." in end_time_str else parse_srt_time(end_time_str)

            chapters.append(Chapter(
                id=len(chapters) + 1,
                title=title,
                start_time=start_time,
                end_time=end_time
            ))
        except (ValueError, IndexError):
            continue

    return chapters


def parse_chapters_file(file_path: str) -> List[Chapter]:
    path = Path(file_path)
    content = path.read_text(encoding="utf-8")
    return parse_chapters_csv(content)


def parse_rules_csv(content: str) -> List[SensitiveRule]:
    rules: List[SensitiveRule] = []
    lines = content.split("\n")
    header_skipped = False

    for line in lines:
        line = line.strip()
        if not line:
            continue

        if not header_skipped:
            header_skipped = True
            continue

        parts = line.split(",", 3)
        if len(parts) < 3:
            continue

        try:
            pattern = parts[0].strip()
            category = parts[1].strip()
            description = parts[2].strip() if len(parts) > 2 else ""
            mask_template = parts[3].strip() if len(parts) > 3 else "[{category}_{index}]"

            rules.append(SensitiveRule(
                id=len(rules) + 1,
                pattern=pattern,
                category=category,
                description=description,
                mask_template=mask_template
            ))
        except (ValueError, IndexError):
            continue

    return rules


def parse_rules_file(file_path: str) -> List[SensitiveRule]:
    path = Path(file_path)
    content = path.read_text(encoding="utf-8")
    return parse_rules_csv(content)


def format_srt_time(td: timedelta) -> str:
    total_seconds = int(td.total_seconds())
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60
    milliseconds = int(td.microseconds / 1000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d},{milliseconds:03d}"


def format_vtt_time(td: timedelta) -> str:
    total_seconds = int(td.total_seconds())
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60
    milliseconds = int(td.microseconds / 1000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d}.{milliseconds:03d}"


def export_srt(subtitles: List[Subtitle]) -> str:
    lines = []
    for sub in subtitles:
        lines.append(str(sub.id))
        lines.append(f"{format_srt_time(sub.start_time)} --> {format_srt_time(sub.end_time)}")
        lines.append(sub.text)
        lines.append("")
    return "\n".join(lines)


def export_vtt(subtitles: List[Subtitle]) -> str:
    lines = ["WEBVTT", ""]
    for sub in subtitles:
        lines.append(f"{format_vtt_time(sub.start_time)} --> {format_vtt_time(sub.end_time)}")
        lines.append(sub.text)
        lines.append("")
    return "\n".join(lines)

File: zy1074/src/parsers/subtitle-parser.ts
import fs from 'fs';
import path from 'path';
import { SubtitleCue } from '../types';
import { timeToSeconds, secondsToTime } from '../utils/time';

export class SubtitleParseError extends Error {
  constructor(message: string, public lineNumber?: number) {
    super(lineNumber ? `第${lineNumber}行: ${message}` : message);
    this.name = 'SubtitleParseError';
  }
}

export class SrtParser {
  parse(content: string): SubtitleCue[] {
    const cues: SubtitleCue[] = [];
    const lines = content.split(/\r?\n/);
    let i = 0;

    while (i < lines.length) {
      const line = lines[i].trim();
      
      if (line === '') {
        i++;
        continue;
      }

      const idMatch = line.match(/^(\d+)$/);
      if (!idMatch) {
        i++;
        continue;
      }

      const id = idMatch[1];
      i++;

      while (i < lines.length && lines[i].trim() === '') {
        i++;
      }

      if (i >= lines.length) {
        throw new SubtitleParseError(`字幕 ${id} 缺少时间轴`, i);
      }

      const timeLine = lines[i].trim();
      const timeMatch = timeLine.match(/^(\d{2}:\d{2}:\d{2}[,.]\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2}[,.]\d{3})/);
      
      if (!timeMatch) {
        throw new SubtitleParseError(`字幕 ${id} 时间轴格式错误: ${timeLine}`, i + 1);
      }

      const startTimeStr = timeMatch[1].replace(',', ',');
      const endTimeStr = timeMatch[2].replace(',', ',');
      
      i++;

      let textLines: string[] = [];
      while (i < lines.length && lines[i].trim() !== '') {
        textLines.push(lines[i]);
        i++;
      }

      const text = textLines.join('\n').trim();

      if (!text) {
        throw new SubtitleParseError(`字幕 ${id} 缺少字幕内容`, i);
      }

      try {
        const startTime = timeToSeconds(startTimeStr);
        const endTime = timeToSeconds(endTimeStr);

        if (startTime >= endTime) {
          throw new SubtitleParseError(`字幕 ${id} 开始时间大于等于结束时间: ${startTimeStr} --> ${endTimeStr}`);
        }

        cues.push({
          id,
          startTime,
          endTime,
          startTimeStr: secondsToTime(startTime, 'srt'),
          endTimeStr: secondsToTime(endTime, 'srt'),
          text
        });
      } catch (error) {
        if (error instanceof SubtitleParseError) {
          throw error;
        }
        throw new SubtitleParseError(`字幕 ${id} 时间解析失败: ${error instanceof Error ? error.message : '未知错误'}`);
      }
    }

    if (cues.length === 0) {
      throw new SubtitleParseError('未找到任何字幕条目');
    }

    return cues;
  }

  parseFile(filePath: string): SubtitleCue[] {
    if (!fs.existsSync(filePath)) {
      throw new SubtitleParseError(`文件不存在: ${filePath}`);
    }

    const ext = path.extname(filePath).toLowerCase();
    if (ext !== '.srt') {
      throw new SubtitleParseError(`不是 SRT 文件: ${filePath}`);
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    return this.parse(content);
  }
}

export class VttParser {
  parse(content: string): SubtitleCue[] {
    const cues: SubtitleCue[] = [];
    const lines = content.split(/\r?\n/);
    let i = 0;

    while (i < lines.length && !lines[i].trim().startsWith('WEBVTT')) {
      i++;
    }

    if (i >= lines.length) {
      throw new SubtitleParseError('缺少 WEBVTT 文件头');
    }

    i++;

    let cueId = 1;
    let inCue = false;
    let currentId: string | null = null;
    let startTimeStr: string | null = null;
    let endTimeStr: string | null = null;
    let textLines: string[] = [];

    while (i < lines.length) {
      const line = lines[i];
      const trimmedLine = line.trim();

      if (trimmedLine === '') {
        if (inCue && startTimeStr && endTimeStr) {
          try {
            const startTime = timeToSeconds(startTimeStr);
            const endTime = timeToSeconds(endTimeStr);

            if (startTime >= endTime) {
              throw new SubtitleParseError(`字幕 ${currentId || cueId} 开始时间大于等于结束时间: ${startTimeStr} --> ${endTimeStr}`);
            }

            const text = textLines.join('\n').trim();
            if (text) {
              cues.push({
                id: currentId || String(cueId),
                startTime,
                endTime,
                startTimeStr: secondsToTime(startTime, 'vtt'),
                endTimeStr: secondsToTime(endTime, 'vtt'),
                text
              });
            }
          } catch (error) {
            if (error instanceof SubtitleParseError) {
              throw error;
            }
            throw new SubtitleParseError(`字幕 ${currentId || cueId} 时间解析失败`);
          }
        }

        inCue = false;
        currentId = null;
        startTimeStr = null;
        endTimeStr = null;
        textLines = [];
        i++;
        continue;
      }

      if (trimmedLine.startsWith('NOTE ') || trimmedLine === 'NOTE') {
        while (i < lines.length && lines[i].trim() !== '') {
          i++;
        }
        continue;
      }

      if (trimmedLine.includes('-->')) {
        inCue = true;
        const timeMatch = trimmedLine.match(/^(\d{2}:\d{2}[:.]\d{2}[,.]\d{3}|\d{2}:\d{2}[,.]\d{3})\s*-->\s*(\d{2}:\d{2}[:.]\d{2}[,.]\d{3}|\d{2}:\d{2}[,.]\d{3})/);
        
        if (!timeMatch) {
          throw new SubtitleParseError(`时间轴格式错误: ${trimmedLine}`, i + 1);
        }

        startTimeStr = timeMatch[1];
        endTimeStr = timeMatch[2];
        cueId++;
        i++;
        continue;
      }

      if (!inCue && trimmedLine.match(/^\w+$/)) {
        currentId = trimmedLine;
        i++;
        continue;
      }

      if (inCue) {
        textLines.push(line);
      }

      i++;
    }

    if (inCue && startTimeStr && endTimeStr) {
      try {
        const startTime = timeToSeconds(startTimeStr);
        const endTime = timeToSeconds(endTimeStr);
        const text = textLines.join('\n').trim();
        
        if (text) {
          cues.push({
            id: currentId || String(cueId),
            startTime,
            endTime,
            startTimeStr: secondsToTime(startTime, 'vtt'),
            endTimeStr: secondsToTime(endTime, 'vtt'),
            text
          });
        }
      } catch (error) {
        if (!(error instanceof SubtitleParseError)) {
          throw new SubtitleParseError(`字幕 ${currentId || cueId} 时间解析失败`);
        }
      }
    }

    if (cues.length === 0) {
      throw new SubtitleParseError('未找到任何字幕条目');
    }

    return cues;
  }

  parseFile(filePath: string): SubtitleCue[] {
    if (!fs.existsSync(filePath)) {
      throw new SubtitleParseError(`文件不存在: ${filePath}`);
    }

    const ext = path.extname(filePath).toLowerCase();
    if (ext !== '.vtt') {
      throw new SubtitleParseError(`不是 VTT 文件: ${filePath}`);
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    return this.parse(content);
  }
}

export function parseSubtitleFile(filePath: string): SubtitleCue[] {
  const ext = path.extname(filePath).toLowerCase();
  
  if (ext === '.srt') {
    const parser = new SrtParser();
    return parser.parseFile(filePath);
  } else if (ext === '.vtt') {
    const parser = new VttParser();
    return parser.parseFile(filePath);
  } else {
    throw new SubtitleParseError(`不支持的字幕格式: ${ext}`);
  }
}
